pycbc/distributions/sky_location.py

```diff
@@ -306,91 +224,115 @@ def from_config(cls, cp, section, variable_args):
-            """Start from a uniform distribution of points, and accepts those
-             whose values on the map are greater than a random value
-             following a uniform law
-
-             Parameters
-             ----------
-             healpix_map  : HealpixMap instance
-             size : int
-                 number of points tested by the method
-             boundaries : list -> tuple of 4 floats
-                 delta_min,delta_max,alpha_min,alpha_max = boundaries
-                 delta is the declination in radians [-pi/2, pi/2]
-                 alpha is the right ascention in radians [0,2pi]
-
-             Returns
-             -------
-            coordinates of the accepted points,
-            following the mhealpy conventions
-
-            """
-            # The angles are in radians and follow the radec convention
-            delta_min, delta_max, alpha_min, alpha_max = boundaries
-
-            # draw points uniformly distributed inside the region delimited by
-            # boundaries
-            u = numpy.random.uniform(0, 1, size)
-            delta = numpy.arcsin(
-                (numpy.sin(delta_max) - numpy.sin(delta_min)) * u
-                + numpy.sin(delta_min)
+        # First of all, draw a random sample of pixel indices following the
+        # given probabilities. This is the trivial part of the algorithm.
+        pix_indices = numpy.random.choice(
+            self.healpix_map.npix, p=self.pix_probs, size=size
+        )
+
+        # Then comes the nasty part of the algorithm. For each selected pixel,
+        # we need to draw a random point uniformly distributed on the patch of
+        # the sky defined by that pixel. We cannot skip points, every pixel
+        # (index) needs precisely one point. We do this via rejection sampling:
+        # we find the (theta,phi) box that encloses the corners of a pixel,
+        # then draw a uniform point inside that box and repeat until the point
+        # falls inside the pixel. The acceptance rate is always around 50% due
+        # to the orientation of the HEALPix pixels. The result seems correct
+        # visually and by checking the acceptance rates. This method is
+        # relatively slow due to the calls to `boundaries()`, which is slow
+        # especially for MOC maps. For MOC maps, I obtain around one million
+        # samples per minute.
+
+        boundaries_vec = self.pixel_corners(pix_indices)
+
+        boundaries_z_min = boundaries_vec[:,2,:].min(axis=1)
+        boundaries_z_max = boundaries_vec[:,2,:].max(axis=1)
+
+        # Find ranges of azimuthal angle. `arctan2()` produces angles in the
+        # range [-π,π] which is not the right convention, so fix that right
+        # away.
+        boundaries_phi = numpy.arctan2(
+            boundaries_vec[:,1,:], boundaries_vec[:,0,:]
+        )
+        del boundaries_vec
+        boundaries_phi[boundaries_phi < 0] += 2 * numpy.pi
```

Comment:
Same question as above.

Reply:
The values here come directly from `numpy.arctan2()`, which returns values in [-π,π], so this is safe (only one "loop" is needed in any case).

I just realized that there is an incorrect edge case here when both arguments of `arctan2()` are negative zeroes, resulting in -π (which should ideally become 0 or -0, but becomes π with the current code). However your proposed solution would not fix this problem either, this would really need a special case.